refactor(ComboBox): Use compound components and context

mayank99 opened the pull request:
Rewrote the entire ComboBox by splitting into compound components and using reducer+context. This provides extra flexibility (e.g. composition) and improves performance (e.g. click lag in virtualization).

One of the important changes I made for performance is save the original index from the passed options into an object whose keys are derived from the option label. This index is also stored as `data-iui-index` in the DOM element. This avoids the need for repeatedly calling `find`/`findIndex`.

## Checklist

- [ ] Add meaningful unit tests for your component (verify that all lines are covered)
- [x] Verify that all existing tests pass
- [x] Add component features demo in Storybook (different stories)
- [x] ~~Approve test images for new stories~~
- [x] ~~Add screenshots of the key elements of the component~~

mayank99 committed:
Finish initial implementation

mayank99 committed:
Fix some (but not all) tests

mayank99 commented on src/core/ComboBox/ComboBox.test.tsx:
```diff
@@ -319,26 +339,28 @@ it('should work with custom itemRenderer', () => {
   ).toHaveTextContent('CUSTOM Item 1');
 
-  fireEvent.keyDown(input, { key: 'ArrowDown' });
+  userEvent.keyboard('{ArrowDown}');
+
   expect(
     document.querySelector('.iui-menu-item.iui-focused.my-custom-item'),
   ).toHaveTextContent('CUSTOM Item 2');
```
I don't understand what is up with these keyboard navigation tests, I can't get them to pass.

mayank99 commented:
@veekeys I think this PR is ready to be hooked up to virtualization. I tried to test it with the current (released) version of `VirtualScroll`, and the performance is great! 👍 

There are two non-performance issues:
- some size miscalculation (which #539 should fix I think)
- in my current keyboard navigation, when pressing ⬇️  on the last item (or ⬆️  on the first), it should loop over. This is not compatible with virtualization so we could try to fix it, or remove this behavior.

mayank99 committed:
Fix menu not opening when typing

mayank99 committed:
Add async example

bentleyvk commented on src/core/ComboBox/ComboBox.tsx:
```diff
@@ -117,252 +122,312 @@ export const ComboBox = <T,>(props: ComboBoxProps<T>) => {
+          })
+        : null;
+
+      return selectedIndex === __originalIndex ||
```
Can you change to normal `if` condition. For me it is hard to read several ternary conditionals in one place.

mayank99 replied:
Removed nested ternary, take a look

bentleyvk commented on src/core/ComboBox/ComboBox.tsx:
```diff
@@ -117,252 +122,312 @@ export const ComboBox = <T,>(props: ComboBoxProps<T>) => {
+  if (!value) {
+    throw new Error(`${context.displayName} is undefined`);
+  }
+  return value!; // eslint-disable-line @typescript-eslint/no-non-null-assertion -- we already checked for undefined
```
Hm, I am not getting this linter error

bentleyvk replied:
Oh getting error later. Strange...

bentleyvk commented on src/core/ComboBox/ComboBox.tsx:
```diff
@@ -117,252 +122,312 @@ export const ComboBox = <T,>(props: ComboBoxProps<T>) => {
+  return value!; // eslint-disable-line @typescript-eslint/no-non-null-assertion -- we already checked for undefined
+};
+
+const ComboBoxInputContainer = (
```
What if you merge it with Input? Is it needed to be as a separate component?

mayank99 replied:
That's how I had it before but this is more flexible. `ComboBox.Input` contains the `<input>` element and all the logic, so for customizing the styling (e.g. icon on the left) you could replace `ComboBox.InputContainer` with something else or even remove it.

bentleyvk commented on src/core/ComboBox/ComboBox.tsx:
```diff
@@ -117,252 +122,312 @@ export const ComboBox = <T,>(props: ComboBoxProps<T>) => {
+            {...rest}
+            isSelected={selectedIndex === __originalIndex}
+            isFocused={focusedIndex === __originalIndex}
+            onClick={() => dispatch(['select', __originalIndex])}
```
Click is still laggy.

mayank99 replied:
Doesn't lag when virtualized. I'm not sure how to fix it in regular mode when there's a large number of items. In my `dispatch(['select'])` I'm setting `isOpen: false` but that triggers a whole bunch of updates before the popover actually closes.

bentleyvk replied:
I think the problem here is that you change `selectedIndex` and `value` is changed from outside that causes full rerender. But if with virtual scroll it is working fine, maybe no need to dig deeper.

mayank99 committed:
forwardref and displayNames

mayank99 committed:
Improve custom async story

mayank99 commented:
So there is an issue with `Disabled Items` story not loading in dev storybook. I ran prod version (`yarn build-storybook`) and also tried in a test app and it loads fine there. That's where I also noticed that keyboard navigation for disabled items seems broken, so I will need to fix that.

I'm thinking of reworking keyboard navigation because it also doesn't work in custom `Async` story and in tests.

mayank99 committed:
Rework keyboard navigation

mayank99 committed:
Fix story

veekeys commented on src/core/ComboBox/ComboBox.tsx:
```diff
@@ -117,252 +123,317 @@ export const ComboBox = <T,>(props: ComboBoxProps<T>) => {
+  // Latest value of the onChange prop
+  const onChangeProp = React.useRef(onChange);
+  React.useEffect(() => {
+    onChangeProp.current = onChange;
```
whats the reason to save it in ref, instead of using directly?

mayank99 replied:
i think this part is the same as before, it's to avoid requiring user to memoize onChange

veekeys commented on src/core/ComboBox/ComboBox.tsx:
```diff
@@ -117,252 +123,317 @@ export const ComboBox = <T,>(props: ComboBoxProps<T>) => {
+    {
+      isOpen: false,
+      selectedIndex: -1,
+      focusedIndex: -1,
```
ammm... cant you save this info instead of looking for index everytime?
Wont now happen `select` event twice? Once from us (lets say keybaord nav) and then once from user (because we will call `onChange` callback and user would change `valueProp` then)

mayank99 replied:
This comment might be in the wrong place, but even if it wasn't, I don't think I understand what you mean by "save this info". Maybe elaborate how I can avoid the lookup?

veekeys commented on src/core/ComboBox/ComboBox.tsx:
```diff
@@ -117,252 +123,317 @@ export const ComboBox = <T,>(props: ComboBoxProps<T>) => {
 
   // Set min-width of menu to be same as input
-  const [minWidth, setMinWidth] = React.useState(0);
+  const [minWidth, setMinWidth] = React.useState(300);
```
remember why needed initial value?

mayank99 replied:
This is leftover from testing, reverted.

veekeys commented on src/core/ComboBox/ComboBox.tsx:
```diff
@@ -117,252 +123,317 @@ export const ComboBox = <T,>(props: ComboBoxProps<T>) => {
-    }
+    dispatch([
+      'select',
+      options.findIndex((option) => option.value === valueProp),
```
yeah, so if you saved initially index info, wouldnt need to look for it here

veekeys commented on src/core/ComboBox/ComboBox.tsx:
```diff
@@ -117,252 +123,317 @@ export const ComboBox = <T,>(props: ComboBoxProps<T>) => {
+
+      return customItem ? (
+        React.cloneElement(customItem, {
+          onClick: (e: unknown) => {
```
It mus be `li` element though, no?

mayank99 replied:
it should be `li` but it can be anything. we dont control it

veekeys replied:
but wouldnt just `React.MouseEvent<HTMLElement>` work here?

mayank99 replied:
`MenuItem` doesn't expose the mouse event

veekeys replied:
Huh? I dont understand. Click even is gonna be either `React.MouseEvent<HTMLElement>` or then will be no event at all (if `MenuItem` does not expose event at all).

mayank99 replied:
no, `MenuItem` (confusingly) passes the `value` of the item.

https://github.com/iTwin/iTwinUI-react/blob/4c9b4a8ae9973303c1ea9314de6958773308b8be/packages/iTwinUI-react/src/core/Menu/MenuItem.tsx#L28-L35
https://github.com/iTwin/iTwinUI-react/blob/4c9b4a8ae9973303c1ea9314de6958773308b8be/packages/iTwinUI-react/src/core/Menu/MenuItem.tsx#L146

veekeys commented on src/core/ComboBox/ComboBox.tsx:
```diff
@@ -117,252 +123,317 @@ export const ComboBox = <T,>(props: ComboBoxProps<T>) => {
+          },
+          // ComboBox.MenuItem handles data-iui-index and iui-focused through context
+          // but we still need to pass them here for backwards compatibility with MenuItem
+          className: cx(customItem.props.className, 'iui-focused'),
```
you need to add `iui-focused` to all elements? doesnt make sense

mayank99 replied:
i have an early return above
<img width="509" alt="image" src="https://user-images.githubusercontent.com/9084735/163568680-0a7648e6-6a85-4a34-8989-6576f04fb25c.png">

bentleyvk commented on src/core/ComboBox/ComboBox.tsx:
```diff
@@ -117,252 +123,317 @@ export const ComboBox = <T,>(props: ComboBoxProps<T>) => {
-    setInputValue(selectedOption?.label ?? '');
-    setFocusedIndex(selectedOption ? options.indexOf(selectedOption) : -1);
-  }, [selectedValue, options]);
+    dispatch(['focus']);
```
Can you move this call to `handleInput`?

mayank99 replied:
Why? `filteredOptions` could also change if `optionsProp` changes for some reason

veekeys reviewed (commented):
Submitting code review. It is pretty complex structure, Im still left in thinking, what and how could be improved :D
But its good as testing with VS went pretty successful.

mayank99 commented on src/core/ComboBox/ComboBox.tsx:
```diff
@@ -380,84 +446,309 @@ export const ComboBox = <T,>(props: ComboBoxProps<T>) => {
+
+const ComboBoxInput = React.forwardRef(
+  (props: InputProps, forwardedRef: React.RefObject<HTMLInputElement>) => {
+    const { onKeyDown: onKeyDownProp, onFocus: onFocusProp, ...rest } = props;
```
this is just to make it clear because i have my own handlers for keydown and focus in this component

mayank99 commented on src/core/ComboBox/ComboBox.tsx:
```diff
@@ -380,84 +446,309 @@ export const ComboBox = <T,>(props: ComboBoxProps<T>) => {
+
+            while (true) {
+              const currentElement =
+                menuRef.current?.querySelector(
```
this is outdated, please look at https://github.com/iTwin/iTwinUI-react/pull/623/commits/69056c9770d96b6bdc0233ded4ce3fca26e384da

but it's still querying dom. can you elaborate on the map part and how it would work with custom items?

mayank99 commented on src/core/ComboBox/ComboBox.tsx:
```diff
@@ -380,84 +446,309 @@ export const ComboBox = <T,>(props: ComboBoxProps<T>) => {
+
+              return dispatch(['focus', Number(prevIndex)]);
+            }
+          } else if (event.key === 'Enter') {
```
space is a typeable character in input

mayank99 committed:
Add displayName to `ComboBoxRefsContext`

mayank99 committed:
Merge onClick handler in `ComboBoxEndIcon`

mayank99 committed:
Use switch for keydown handler

mayank99 committed:
Reset minWidth to 0

mayank99 commented on src/core/ComboBox/ComboBox.tsx:
```diff
@@ -380,84 +446,309 @@ export const ComboBox = <T,>(props: ComboBoxProps<T>) => {
+
+const ComboBoxPopover = React.forwardRef(
+  (
+    props: PopoverProps & { children: JSX.Element },
```
what for? For me it's nicer if we use `children` for composing those subcomponents.

veekeys replied:
well.. just if we have `Popover` and then we have `ComboBoxPopover` kinda implies we talk about same `Popover` and having consistent APIs somehow feels better.

mayank99 commented on src/core/ComboBox/ComboBox.tsx:
```diff
@@ -380,84 +446,309 @@ export const ComboBox = <T,>(props: ComboBoxProps<T>) => {
+          const length =
+            menuRef.current?.querySelectorAll('[data-iui-index]').length ?? 0;
+
+          if (event.key === 'ArrowDown') {
```
no reason. using `switch` now in latest version

mayank99 commented on src/core/ComboBox/ComboBox.tsx:
```diff
@@ -380,84 +446,309 @@ export const ComboBox = <T,>(props: ComboBoxProps<T>) => {
+      <span
+        ref={refs}
+        className={cx('iui-end-icon', className)}
+        onClick={() => dispatch(['toggle'])}
```
done

mayank99 commented on src/core/ComboBox/ComboBox.tsx:
```diff
@@ -117,252 +123,312 @@ export const ComboBox = <T,>(props: ComboBoxProps<T>) => {
-          },
-        });
-      }
+const ComboBoxRefsContext = React.createContext<
```
added displayName

mayank99 committed:
Reset input value when dropdown is closed

bentleyvk commented:
Why tests are still failing?

veekeys commented on src/core/ComboBox/ComboBox.tsx:
```diff
@@ -180,189 +174,274 @@ export const ComboBox = <T,>(props: ComboBoxProps<T>) => {
+          key={option.id ?? __originalIndex}
+          {...rest}
+          isSelected={selectedIndex === __originalIndex}
+          onClick={() => dispatch(['select', __originalIndex])}
```
onclick might need merge here too

mayank99 replied:
with what?

veekeys replied:
`option.onClick`

veekeys commented on src/core/ComboBox/ComboBox.tsx:
```diff
@@ -180,189 +174,274 @@ export const ComboBox = <T,>(props: ComboBoxProps<T>) => {
+  if (!value) {
+    throw new Error(`${context.displayName} is undefined`);
+  }
+  return value!; // eslint-disable-line @typescript-eslint/no-non-null-assertion -- we already checked for undefined
```
this is weird.. why it fails to see the check above? because no return?
Maybe this whole function can be moved to utils and reused, right?

veekeys commented on src/core/ComboBox/ComboBox.tsx:
```diff
@@ -380,84 +459,365 @@ export const ComboBox = <T,>(props: ComboBoxProps<T>) => {
+        filtered: optionsRef.current.findIndex(
+          (el) => Number(el.dataset.iuiIndex) === focusedIndex,
+        ),
+      };
```
`optionsRef` in deps will do nothing.

mayank99 replied:
Yeah, but the linter needs it because it's coming from context.

veekeys commented on src/core/ComboBox/ComboBox.tsx:
```diff
@@ -380,84 +459,365 @@ export const ComboBox = <T,>(props: ComboBoxProps<T>) => {
+);
+ComboBoxMenu.displayName = 'ComboBoxMenu';
+
+const ComboBoxMenuItem = React.memo(
```
Why memo?

mayank99 replied:
To avoid excessive rerenders. Is there a downside to not having it?

veekeys replied:
oh, if you checked the performance gain, thats good.

bentleyvk commented on src/core/ComboBox/ComboBox.tsx:
```diff
@@ -180,189 +174,274 @@ export const ComboBox = <T,>(props: ComboBoxProps<T>) => {
+      if (value === valueProp) {
+        return;
+      }
+      onChangeProp.current?.(value);
```
It will be also called on options change and on selection, not only on value change.

bentleyvk commented on src/core/ComboBox/ComboBox.tsx:
```diff
@@ -180,189 +174,274 @@ export const ComboBox = <T,>(props: ComboBoxProps<T>) => {
+            index: __originalIndex,
+            id: option.id,
+          })
+        : null;
```
Here could be `<ComboboxMenuItem>`. That way no need another trinary on return.

bentleyvk commented on src/core/ComboBox/ComboBox.tsx:
```diff
@@ -180,189 +174,274 @@ export const ComboBox = <T,>(props: ComboBoxProps<T>) => {
-        </MenuExtraContent>,
-      ];
+  return (
+    <ComboBoxRefsContext.Provider
```
Why 3 providers instead of 1?

mayank99 replied:
when any value inside a context changes, it rerenders the whole component that uses the context, even if that component is not using the part that changed. So it is better to split into multiple contexts.

bentleyvk replied:
I know that it causes rerenders? But is it a case here?

bentleyvk commented on src/core/ComboBox/ComboBox.tsx:
```diff
@@ -180,189 +174,274 @@ export const ComboBox = <T,>(props: ComboBoxProps<T>) => {
+                endIcon={
+                  <ComboBoxEndIcon
+                    className={cx({
+                      'iui-actionable': !inputProps?.disabled,
```
These could be props and classes could be moved inside icon component.

mayank99 replied:
So you want 3 new props on this super simple component?

mayank99 replied:
Added

bentleyvk commented on src/core/ComboBox/ComboBox.tsx:
```diff
@@ -180,189 +174,274 @@ export const ComboBox = <T,>(props: ComboBoxProps<T>) => {
+            <>
+              <ComboBoxInputContainer
+                disabled={inputProps?.disabled}
+                endIcon={
```
Why not just place icon at the end, instead of having prop?

mayank99 replied:
Using `children` now 👍

bentleyvk commented on src/core/ComboBox/ComboBox.tsx:
```diff
@@ -180,189 +174,274 @@ export const ComboBox = <T,>(props: ComboBoxProps<T>) => {
+      return {
+        ...state,
+        isOpen: false,
+        selectedIndex: value ?? state.focusedIndex ?? state.selectedIndex,
```
How can selected index be focused index?

mayank99 replied:
it selects the currently focused index if value is not passed in the dispatch

bentleyvk replied:
Does it make sense to select focused item? I would say it should be the other way arround.

mayank99 replied:
I have both cases.
![image](https://user-images.githubusercontent.com/9084735/165173138-9512afc5-a55d-4791-a293-ee034e3990c1.png)

bentleyvk commented on src/core/ComboBox/ComboBox.tsx:
```diff
@@ -380,84 +459,365 @@ export const ComboBox = <T,>(props: ComboBoxProps<T>) => {
-            </Menu>
+
+            let nextIndex = focusedIndexRef.current.filtered;
+            do {
```
Can you just get all focusable items and increase index of focusable list?

mayank99 replied:
getFocusableElements will not work with virtualized lists because the elements are not always in the DOM

bentleyvk commented on src/core/ComboBox/ComboBox.tsx:
```diff
@@ -380,84 +459,365 @@ export const ComboBox = <T,>(props: ComboBoxProps<T>) => {
+
+    React.useEffect(() => {
+      if (menuRef.current != null) {
+        const items = menuRef.current.querySelectorAll('[data-iui-index]');
```
What won't work in custom rendered items that are missing this attribute?

mayank99 replied:
keyboard navigation

mayank99 replied:
and how would i get original index from children if this attribute is not present?

bentleyvk commented on src/core/ComboBox/ComboBox.tsx:
```diff
@@ -180,189 +174,274 @@ export const ComboBox = <T,>(props: ComboBoxProps<T>) => {
 
-  // update inputValue and focusedIndex every time selected value changes
+  // Reset focused item when filter changes
   React.useEffect(() => {
```
Can it be that this reset is causing you a bug where initialy selected item is not focused?

mayank99 replied:
I tried to remove it and it didn't fix the bug

bentleyvk commented on stories/core/ComboBox.stories.tsx:
```diff
@@ -497,3 +503,104 @@ WithCustomMessageIcon.args = {
+  };
+
+  return (
+    <ComboBox options={[]}>
```
Isn't this demo too complex?

veekeys reviewed (commented):
I still need to play around with it and check up keyboard nav

mayank99 commented on src/core/ComboBox/ComboBox.tsx:
```diff
@@ -14,10 +14,9 @@ import {
+  useMergedRefs,
 } from '../utils';
 import SvgCaretDownSmall from '@itwin/itwinui-icons-react/cjs/icons/CaretDownSmall';
 import 'tippy.js/animations/shift-away.css';
```
It's the same as before

mayank99 committed:
Move `useSafeContext` to utils

mayank99 committed:
Update comment

mayank99 commented on src/core/ComboBox/ComboBox.tsx:
```diff
@@ -180,189 +174,274 @@ export const ComboBox = <T,>(props: ComboBoxProps<T>) => {
+        <ComboBoxMenuItem
+          key={(__originalIndex as number) ?? index}
+          index={option.__originalIndex}
+          onClick={() => dispatch(['select', __originalIndex as number])}
```
that's what the `rest` in next line does

veekeys replied:
well, I meant the `key`, `onClick`, etc props, which you assign yourself.

mayank99 replied:
The option type doesn't have any of those props, so is it expected that the user will pass something in there? The arbitrary `[key]` is mainly for `aria-` and `data-` attributes iirc.
![image](https://user-images.githubusercontent.com/9084735/166316060-dd888a9c-5c65-4122-b7ad-42a9d47fffbb.png)

mayank99 committed:
Prefer `option.id` in extended options

mayank99 commented on src/core/ComboBox/ComboBox.tsx:
```diff
@@ -117,60 +123,48 @@ export const ComboBox = <T,>(props: ComboBoxProps<T>) => {
+    setOptions(
+      optionsProp.map((option, index) => ({
+        ...option,
+        id: `${id}-option${index}`,
```
👍 https://github.com/iTwin/iTwinUI-react/pull/623/commits/666c0686bd6acce4c1c60dfb974301fa6813ed76

mayank99 committed:
Fix keyboard navigation

mayank99 committed:
Merge branch 'main' into mayank/combobox-refactor

mayank99 commented:
I've reworked keyboard navigation (again), kinda reverting my last rework. Now that we are ok with losing focused index when using virtual scroll + mouse, It seems to work pretty well in all existing cases (regular, controlled, custom renderer, custom composed).

While testing custom composed (`Async` story) example, I noticed that pressing <kbd>⏎</kbd> and <kbd>Esc</kbd> does not work, which is kinda expected because there are no `options` available in state. I'm thinking we might want to provide something like an `onHighlight` callback (which will be called every time a menu item is focused) so that the user can handle it on their end.

mayank99 commented on src/core/ComboBox/ComboBox.tsx:
```diff
@@ -180,189 +174,274 @@ export const ComboBox = <T,>(props: ComboBoxProps<T>) => {
+        })
+      ) : (
+        <ComboBoxMenuItem
+          key={(__originalIndex as number) ?? index}
```
hmm. I think there was a reason I added this, but can't remember.

mayank99 committed:
Remove casts from `__originalIndex`

mayank99 commented on src/core/ComboBox/ComboBox.tsx:
```diff
@@ -180,189 +174,274 @@ export const ComboBox = <T,>(props: ComboBoxProps<T>) => {
+      return customItem ? (
+        React.cloneElement(customItem, {
+          onClick: (e: unknown) => {
+            dispatch(['select', __originalIndex as number]);
```
Apparently not. Removed in [`f9e42d5`](https://github.com/iTwin/iTwinUI-react/pull/623/commits/f9e42d55b02ce8a225fbbd0765c9830d61300276)

bentleyvk commented:
> While testing custom composed (`Async` story) example, I noticed that pressing ⏎ and Esc does not work, which is kinda expected because there are no `options` available in state. I'm thinking we might want to provide something like an `onHighlight` callback (which will be called every time a menu item is focused) so that the user can handle it on their end.

Can you use `itemRenderer` instead of custom popover?

bentleyvk commented on src/core/ComboBox/ComboBox.tsx:
```diff
@@ -180,189 +175,268 @@ export const ComboBox = <T,>(props: ComboBoxProps<T>) => {
+
+  // Memoize the default (not selected/focused) version of menu items
+  const memoizedItems = React.useMemo(() => {
+    return options.map((option, index) => {
```
index is not used

mayank99 replied:
removed

bentleyvk commented:
Is it only for me, that after selecting a value, I can't open the combobox menu?

veekeys commented on stories/core/ComboBox.stories.tsx:
```diff
@@ -424,8 +424,14 @@ export const CustomRenderer: Story<Partial<ComboBoxProps<string>>> = (args) => {
-    ({ value, label }, { isSelected, id }) => (
-      <MenuItem key={value} id={id} isSelected={isSelected} value={value}>
+    ({ value, label }, { isSelected, id, index }) => (
+      <ComboBox.MenuItem
```
Would be good to leave story as it was so we are sure nothing got broken using it "the old way"

mayank99 replied:
Reverted story back to use `MenuItem`. Found a bug with scrollIntoView for which I had to add a ref in the `cloneElement`

bentleyvk commented on src/core/ComboBox/ComboBox.tsx:
```diff
@@ -180,189 +175,268 @@ export const ComboBox = <T,>(props: ComboBoxProps<T>) => {
-      return;
-    }
+    dispatch([
+      'select',
```
Don't you need to set new focus index too?

veekeys commented on src/core/ComboBox/ComboBox.tsx:
```diff
@@ -301,7 +301,9 @@ export const ComboBox = <T,>(props: ComboBoxProps<T>) => {
           // but we still need to pass them here for backwards compatibility with MenuItem
-          className: cx(customItem.props.className, 'iui-focused'),
+          className: cx(customItem.props.className, {
+            'iui-focused': focusedIndex === __originalIndex,
```
now this got weird :D you said yourself, that you have early return, if focusedIndex would equal to original one

mayank99 replied:
i've read this comment about 10 different times in 4 days and i'm still not sure what it means 😄

veekeys replied:
Oh, I think previously there wasnt (or I didnt notice) check for selectedIndex too.
So when I commented, idea was:
You already check if focused index is not equal to original and early return. So on that particalar code line `focused === __originalIndex` could be replaced just with `true` as it was before.

veekeys commented on src/core/ComboBox/ComboBox.tsx:
```diff
@@ -545,62 +522,73 @@ const ComboBoxInput = React.forwardRef(
             event.preventDefault();
             if (!isOpen) {
-              return dispatch(['open']);
+              dispatch(['open']);
```
I'd say, if you open, just focus first item too.

mayank99 replied:
there is no `return`, so it will perform the appropriate action after opening

veekeys replied:
I dont think it fully performs the appropiate action now.
![image](https://user-images.githubusercontent.com/81580355/165279848-cff0ee4e-6eae-4d23-b333-5c21067fbd9f.png)

After selected value and opening again with keyboard, it does not focus selected value.

bentleyvk commented on src/core/ComboBox/ComboBox.tsx:
```diff
@@ -380,84 +454,363 @@ export const ComboBox = <T,>(props: ComboBoxProps<T>) => {
+              );
+              const nextElement =
+                currentElement?.nextElementSibling ??
+                menuRef.current?.querySelector('[data-iui-index]');
```
When it can't have sibling?

mayank99 replied:
when at the first or last item

bentleyvk commented on src/core/ComboBox/ComboBox.tsx:
```diff
@@ -380,84 +454,363 @@ export const ComboBox = <T,>(props: ComboBoxProps<T>) => {
+              if (nextElement?.ariaDisabled !== 'true') {
+                return dispatch(['focus', __originalIndex]);
+              }
+            } while (nextIndex !== focusedIndexRef.current);
```
Won't it get into infinite loop with 1 item?

mayank99 replied:
i didn't find any infinite loop

veekeys commented on src/core/ComboBox/ComboBox.tsx:
```diff
@@ -380,84 +454,363 @@ export const ComboBox = <T,>(props: ComboBoxProps<T>) => {
+                currentElement?.nextElementSibling ??
+                menuRef.current?.querySelector('[data-iui-index]');
+              nextIndex = Number(nextElement?.getAttribute('data-iui-index'));
+              const __originalIndex = Number(
```
amm.. why you need `__originalIndex` ? Value same as `nextIndex` just with -1 if get undefined attribute.

mayank99 replied:
removed

bentleyvk commented on src/core/ComboBox/ComboBox.tsx:
```diff
@@ -380,84 +454,363 @@ export const ComboBox = <T,>(props: ComboBoxProps<T>) => {
+                currentElement?.previousElementSibling ??
+                menuRef.current?.querySelector('[data-iui-index]:last-of-type');
+              prevIndex = Number(prevElement?.getAttribute('data-iui-index'));
+              const __originalIndex = Number(
```
Move it under `if` where it is used.

mayank99 replied:
i've removed it

mayank99 committed:
Merge branch 'main' into mayank/combobox-refactor

mayank99 committed:
Fix keyboard navigation (again)

mayank99 committed:
Close menu on tab

mayank99 committed:
Fix tests

mayank99 commented:
> Can you use `itemRenderer` instead of custom popover?

No, because there are no `options`. It's either full composition or nothing.

I really think we should have some useful props like `onSelect` and `onHighlight` so that users could build all combinations of things.

mayank99 commented:
> Is it only for me, that after selecting a value, I can't open the combobox menu?

This should be fixed now. All tests are passing too.

veekeys commented on packages/iTwinUI-react/src/core/ComboBox/ComboBox.tsx:
```diff
@@ -382,84 +457,349 @@ export const ComboBox = <T,>(props: ComboBoxProps<T>) => {
+              );
+              const nextElement =
+                currentElement?.nextElementSibling ??
+                menuRef.current?.querySelector('[data-iui-index]');
```
What this would do? Just get first element, which has `data-iui-index`.

veekeys replied:
a ok.. I guess thats the part which stops cycle going infinity

veekeys commented on packages/iTwinUI-react/src/core/ComboBox/ComboBox.tsx:
```diff
@@ -382,84 +457,349 @@ export const ComboBox = <T,>(props: ComboBoxProps<T>) => {
+        const length = optionsRef.current.length ?? 0;
+
+        switch (event.key) {
+          case 'ArrowDown': {
```
hm...whats not nice now, is that if you hold arrow, nothing happens.. You dont see list going down anymore..

bentleyvk replied:
It is going down for me on holding the key. But I for me it is lagging. Opened deployed original version and it is navigating smoothly.

veekeys replied:
Yeah, cause other one for me did not even load (disabled one)

adamhock replied:
Holding down arrow works for me

veekeys replied:
It still does not work for me... :/

veekeys commented on packages/iTwinUI-react/stories/core/ComboBox.stories.tsx:
```diff
@@ -424,8 +424,14 @@ export const CustomRenderer: Story<Partial<ComboBoxProps<string>>> = (args) => {
   }, []);
 
   const itemRenderer = React.useCallback(
```
story controls dissapeared now.

mayank99 replied:
Any idea why they disappeared? I haven't changed anything in most of the stories.

Not that they were useful, but still.

veekeys replied:
I think some of them returned now :/

bentleyvk commented on packages/iTwinUI-react/src/core/ComboBox/ComboBox.tsx:
```diff
@@ -382,84 +457,349 @@ export const ComboBox = <T,>(props: ComboBoxProps<T>) => {
+);
+ComboBoxMenuItem.displayName = 'ComboBoxMenuItem';
+
+ComboBox.InputContainer = ComboBoxInputContainer;
```
Maybe have subcomponents in different files like in other our components. Because it is hard to view ~1k lines of code.

mayank99 replied:
i think @veekeys will disagree

veekeys replied:
Carousel we  did separate in files. So I guess lets be consistent..?

adamhock replied:
I agree to put subcompontents in different files

mayank99 replied:
I've split everything into separate files 👍

veekeys replied:
But you do not export them. On purpose just as internal use?

mayank99 replied:
Yup, lets keep internal for now. We can expose a better API with a better custom example in a future PR

bentleyvk commented on packages/iTwinUI-react/src/core/ComboBox/ComboBox.tsx:
```diff
@@ -382,84 +457,349 @@ export const ComboBox = <T,>(props: ComboBoxProps<T>) => {
+);
+ComboBoxInput.displayName = 'ComboBoxInput';
+
+const ComboBoxPopover = React.forwardRef(
```
Maybe worth calling it `ComboBoxMenu` or `ComboBoxDropdown`

mayank99 replied:
Renamed to `ComboBoxDropdown`.

There is already a `ComboBoxMenu` expected as a child of this.

bentleyvk commented on packages/iTwinUI-react/stories/core/ComboBox.stories.tsx:
```diff
@@ -9,12 +9,12 @@ import {
 import { CreeveyStoryParams } from 'creevey';
-import { SvgCamera } from '@itwin/itwinui-icons-react';
+import { SvgCamera, SvgSearch } from '@itwin/itwinui-icons-react';
 
```
Disabled story takes a long time to load, whole window freezes and fans start to go wild.

bentleyvk replied:
Any ideas what is the cause?

mayank99 replied:
idk. must be some storybook bug. I don't think it's related to this PR

bentleyvk replied:
But why it was working previously? 😀

mayank99 replied:
it lags for me even without this PR, so I don't think it was working previously

mayank99 commented on packages/iTwinUI-react/src/core/ComboBox/ComboBox.tsx:
```diff
@@ -183,188 +175,271 @@ export const ComboBox = <T,>(props: ComboBoxProps<T>) => {
-      setInputValue(event.target.value);
+      const { value } = event.currentTarget;
+      setInputValue(value);
+      dispatch(['open']);
```
if you select an option, it closes. so it should reopen if closed when typing

adamhock commented on packages/iTwinUI-react/src/core/ComboBox/ComboBox.test.tsx:
```diff
@@ -209,77 +210,99 @@ it('should handle keyboard navigation', () => {
   expect(input).toHaveAttribute('aria-controls', `${id}-list`);
 
-  const items = document.querySelectorAll('.iui-menu-item');
+  let items = document.querySelectorAll('.iui-menu-item');
```
I don't understand this change and why `items` needs to be redefined every time

mayank99 replied:
the test was failing and when I tried debugging, I saw that `items` was referring to the old instance, so it had to be queried again. i have no clue how/why it was working before, but this change was just to get the test to pass.

mayank99 committed:
Typo in args

mayank99 committed:
Add disabled and isOpen props to `ComboBoxEndIcon`

mayank99 committed:
Fix arrow down losing index

mayank99 committed:
Split everything into separate files

mayank99 committed:
Remove subcomponents from barrell

mayank99 committed:
Add back `scrollIntoView` ref for custom items

mayank99 committed:
Merge branch 'main' into mayank/combobox-refactor

mayank99 committed:
Use `children` for ComboBoxEndIcon

mayank99 committed:
Remove `ManyItems` story

mayank99 commented:
I've split everything into separate files and removed the subcomponents from barrel. I've also removed the `Async` story (we should add this in a future PR when we have a better API) and the `Many Items` story (we should add this in the virtualization PR).

veekeys commented:
Can we switch back behavior, that when you are at the last item, you cannot appear on the first?
Just in case Ill get issues with VS to achieve this :D

veekeys commented on packages/iTwinUI-react/src/core/ComboBox/ComboBoxDropdown.tsx:
```diff
@@ -0,0 +1,47 @@
+      <Popover
+        placement='bottom-start'
+        visible={isOpen}
+        onClickOutside={React.useCallback(
```
using inline hooks behaves same as if you had variable, right?

veekeys commented on packages/iTwinUI-react/src/core/ComboBox/ComboBoxMenu.tsx:
```diff
@@ -0,0 +1,46 @@
+    return (
+      <Menu
+        id={`${id}-list`}
+        style={React.useMemo(
```
Inline hooks act same as if you had them as variables?

mayank99 replied:
well yeah

mayank99 commented:
> Can we switch back behavior, that when you are at the last item, you cannot appear on the first? Just in case Ill get issues with VS to achieve this :D

But it's nice to have this feature. We can always disable this for VS if it causes issues.

mayank99 committed:
Remove memoized items

mayank99 committed:
Explicitly select focusedIndex

mayank99 committed:
Fix options not resetting

mayank99 committed:
Fix test

veekeys commented on packages/iTwinUI-react/src/core/ComboBox/ComboBox.tsx:
```diff
@@ -45,9 +52,9 @@ export type ComboBoxProps<T> = {
    */
   filterFunction?: (
-    options: SelectOption<T>[],
+    options: ComboBoxOption<T>[],
```
Isnt this one a breaking change????

mayank99 replied:
Is it? This is a function we call. The user just takes whatever from the `options`.

veekeys replied:
In my understanding it is the change of API (types).
Not only the coming argument type changed (so if users defined that explicitly, will get errors), but also the return type changed (if users do not do `..rest` they will end up with errors too ) :/

mayank99 replied:
Oh yeah, good point about return type. Hmm, I will revert this then

mayank99 replied:
This is going to be tough because our `filteredOptions` state expects `ComboBoxOption` so that it is able to access `__originalIndex`

veekeys replied:
What can we do?

veekeys replied:
Lets fix this and I approve. Too long hanging.

mayank99 replied:
Fixed by saving `__originalIndex` in `optionsExtraInfoRef` and getting rid of `ComboBoxOption`

mayank99 commented on packages/iTwinUI-react/src/core/ComboBox/helpers.ts:
```diff
@@ -0,0 +1,64 @@
+      };
+    }
+    case 'focus': {
+      return { ...state, focusedIndex: value ?? state.selectedIndex ?? -1 };
```
no, we want to fall back to `selectedIndex` because that's what needs to be focused when `value` is not passed

mayank99 committed:
dispatch conditionally in ComboBoxEndIcon

mayank99 committed:
revert import order

mayank99 committed:
Merge branch 'main' into mayank/combobox-refactor

mayank99 commented on packages/iTwinUI-react/src/core/ComboBox/ComboBoxEndIcon.tsx:
```diff
@@ -0,0 +1,51 @@
+        onClick={(e) => {
+          onClickProp?.(e);
+          if (!e.isDefaultPrevented()) {
+            dispatch(['toggle']);
```
um ok, done

adamhock replied:
Now the `"toggle"` case can be removed from the `ComboBoxAction` list in helper.ts

adamhock replied:
One last spot [here](https://github.com/iTwin/iTwinUI-react/pull/623/files#diff-65491c8f7818bdc2149a18e25ea8664c7eb9f092ce4961187bc4f5159baeae9aR3)

mayank99 replied:
removed

mayank99 commented on packages/iTwinUI-react/stories/core/ComboBox.stories.tsx:
```diff
@@ -9,9 +9,9 @@ import {
-  MenuItem,
   StatusMessage,
   SelectOption,
+  MenuItem,
```
done

mayank99 committed:
Remove `'toggle'` action

mayank99 committed:
Replace `ComboBoxOption` with `optionsExtraInfoRef`

mayank99 committed:
Replace `optionsRef` with `optionsExtraInfoRef`

mayank99 committed:
typo in deps array

mayank99 committed:
Remove `children` and update jsdoc

veekeys commented on packages/iTwinUI-react/src/core/ComboBox/ComboBox.tsx:
```diff
@@ -155,6 +143,28 @@ export const ComboBox = <T,>(props: ComboBoxProps<T>) => {
+  }, [id, options]);
+
+  // Initialize the extra info only if it is not already initialized
+  if (
```
Why this approach instead of `useEffect` ?

mayank99 replied:
if using `useEffect`, this ref is not available on first render so I can't access `__originalIndex` inside `getMenuItem`

veekeys replied:
ref is available in useEffect, unless element is conditional.

veekeys replied:
a wait, ref of menu items?

mayank99 replied:
`optionsExtraInfoRef` doesn't get set until after the useEffect, so it's not available when `getMenuItem` is first called

veekeys commented on packages/iTwinUI-react/src/core/ComboBox/ComboBox.tsx:
```diff
@@ -70,8 +70,10 @@ export type ComboBoxProps<T> = {
+   *
+   * For keyboard navigation to work, the returned element should use the `id` provided by this function.
+   * The `isFocused` state is calculated using this `id` and can be used for specifying the focus styling.
+   * If a `MenuItem` is returned, then focus sytling is automatically handled.
```
`sytling` you say.
Where is vscode extension of spell checker????

mayank99 replied:
somehow missed it
![image](https://user-images.githubusercontent.com/9084735/168081333-fc0524cf-7c90-43e5-aedd-f2483f63281d.png)

fixed

veekeys replied:
so bad marking... :O

veekeys commented on packages/iTwinUI-react/src/core/ComboBox/ComboBox.tsx:
```diff
@@ -82,10 +84,6 @@ export type ComboBoxProps<T> = {
-   * Children can be specified to customize individual parts.
-   */
-  children?: React.ReactNode;
 } & Pick<InputContainerProps, 'status'> &
```
why children removed?

bentleyvk commented on packages/iTwinUI-react/src/core/ComboBox/ComboBox.test.tsx:
```diff
@@ -10,6 +10,7 @@ import { SvgCaretDownSmall } from '@itwin/itwinui-icons-react';
 import { MenuItem } from '../Menu';
 import { StatusMessage } from '../StatusMessage';
 import userEvent from '@testing-library/user-event';
+import { act } from 'react-dom/test-utils';
```
Why it is imported not from testing-library?

mayank99 replied:
Ask vscode autoimport. Fixed

mayank99 committed:
tytpo

mayank99 committed:
length - 1

mayank99 committed:
Fix `act` import

veekeys reviewed (approved):
Great job!
Looking forward to bug-free use cases of composition.
